lisa/sut_orchestrator/libvirt/platform.py

```diff
@@ -1341,3 +1388,316 @@ def _capture_libvirt_logs(self) -> None:
+
+        # if not iommu_group:
+        #     raise SkippedException(
+        #         f"Pool {pool_type} running out of devices: {pool}, "
+        #         "No IOMMU Group has sufficient count of devices, "
+        #         f"Refer: {pool}"
+        #     )
+
+        # # Remove the devices and update the pool
+        # # Remove entire iommu group device list so conflict does not happen
+        # devices = pool.pop(iommu_group)
+        # self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+        # return devices
+
+        keys = list(pool.keys())
+        results = []
+        for r in range(1, len(keys) + 1):
+            for combo in combinations(keys, r):
+                if sum(len(pool.get(key, [])) for key in combo) == count:
+                    results.append(combo)
+        if not results:
+            for r in range(1, len(keys) + 1):
+                for combo in combinations(keys, r):
+                    if sum(len(pool.get(key, [])) for key in combo) >= count:
+                        results.append(combo)
+                        break
+                if results:
+                    break
+
+        if not results:
+            raise SkippedException(
+                f"Pool {pool_type} running out of devices: {pool}, "
+                "No IOMMU Group has sufficient count of devices, "
+                f"Refer: {pool}"
+            )
+
+        devices: List[DeviceAddressSchema] = []
+        selected_pools = results[0]
+        for iommu_grp in selected_pools:
+            devices += pool.pop(iommu_grp)
+        self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+        return devices
+
+    def _put_devices_into_pool(
+        self,
+        node_context: NodeContext,
+    ) -> None:
+        device_context = node_context.device_passthrough_context
+        for context in device_context:
+            pool_type = context.pool_type
+            devices_list = context.device_list
+            pool = self.AVAILABLE_HOST_DEVICES.get(pool_type, {})
+            for device in devices_list:
+                iommu_grp = self._get_device_iommu_group(device)
+                pool_devices = pool.get(iommu_grp, [])
+                pool_devices.append(device)
+                pool[iommu_grp] = pool_devices
+            self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+
+    def _verify_device_passthrough_post_boot(
+        self,
+        node_context: NodeContext,
+    ) -> None:
+        device_context = node_context.device_passthrough_context
+        for context in device_context:
+            devices = context.device_list
+            for device in devices:
+                err = f"Kernel driver is not vfio-pci for device: {device}"
+                pool_type = context.pool_type.value
+                if context.managed == "yes" and "pci" in pool_type:
+                    is_vfio_pci = self._is_driver_vfio_pci(device)
+                    assert is_vfio_pci, err
+
+    def _check_passthrough_support(self, node: Node) -> None:
```

Comment:
Might be worth renaming `node` to `host_node`, just to clarify this is running on the host and not a test VM.

Reply:
Sure Chris

Reply:
Done